### PR TITLE
helpers, nfs: recover panics across the NFS server

### PR DESCRIPTION
**Pull request by AliRamberg**
## Summary

`RecoverPanics` (added in #6) wraps the `nfs.Handler` and the `billy.Filesystem` it returns, but **not** the `billy.File` that filesystem hands back. go-nfs calls `ReadAt`/`WriteAt` directly on that file:

- `nfs_onread.go:70` — `fh.ReadAt(resp.Data, int64(obj.Offset))`
- `nfs_onwrite.go:86` — `file.WriteAt(req.Data[:end], int64(req.Offset))`

`recoveryFilesystem.Open/OpenFile/Create/TempFile` return the file **unwrapped**, and the method's own `defer recover()` only covers the duration of the open call — not the later `ReadAt`/`WriteAt`. So a panic in a file operation escapes the recovery wrappers, unwinds the per-connection `serve` goroutine (`conn.go`), and crashes the server. This is exactly the gap that forces downstream consumers (e.g. lakeFS Enterprise / Everest) to keep their own `defer recover()` around every `ReadAt`.

## Changes

- Add a `recoveryFile` that wraps a `billy.File` and recovers panics in every method (`Read`, `Write`, `ReadAt`, `WriteAt`, `Seek`, `Stat`, `Lock`, `Unlock`, `Truncate`, `Close`, `Name`). Like the existing wrappers it uses explicit delegation, not embedding, so new interface methods fail to compile until implemented.
- Wrap the file returned by `Create`/`Open`/`OpenFile`/`TempFile` in `recoveryFile`.
- File operations that return an `error` surface a recovered panic as a new exported sentinel `ErrRecoveredPanic` (via `%w`), instead of returning a zero-value "success" — so a panicking read/write fails the request rather than returning bogus data or a silent 0-byte read.
- Add `TestRecoverPanicsFile` (table-driven, `ReadAt`/`WriteAt`) asserting both that `onPanic` fires and that the call returns `ErrRecoveredPanic`.

## Verification

- `go build ./...` — clean (pre-existing `example/osview` failure is unrelated; that package is gated on memphis not yet supporting go-billy/v6)
- `go vet ./helpers/...` — clean
- `go test -race ./helpers/` — pass
- `gofmt -l` — clean

## Related

- Follows #6 (handler + filesystem recovery)
- Lets downstream consumers drop their per-`ReadAt` panic guards

### CODE REVIEW

**Comment by AliRamberg** [comment]
Pushed three commits extending this beyond `billy.File` to close the remaining unprotected panic paths. Branch was behind master (which added the streaming `DirIterator`), so the first commit is the master merge — this PR couldn't merge without it.

- **`a551f77`** — merge `master`; resolves the `recovery.go` divergence as the union of `recoveryFile` (this PR) and `recoveryHandlerWithDirIterator` (master).
- **`b1e0fbd`** — `helpers`:
  - `recoveryDirIterator` wraps the `nfs.DirIterator` returned by `OpenDir` (`Next`/`FileInfo`/`Cookie`/`Verifier`/`Close`); go-nfs drives these directly while streaming a listing, so wrapping `OpenDir` alone left them exposed. A recovered `Next` panic ends the listing.
  - `recoveryFilesystem` error-returning methods now surface `ErrRecoveredPanic` via `%w` instead of returning a zero-value "success" (e.g. `Open` returning a nil file with a nil error, which the caller dereferences).
- **`779aa2e`** — `nfs`: `conn.dispatch` recovers panics around the handler call in the per-connection serve goroutine. This is the only layer that catches panics in handler dispatch / XDR decode that no wrapper can reach. A recovered panic becomes `SERVERFAULT` → an NFS error reply when no response has been written; the connection stays up. (Edge case: a panic *after* a partial response has been written still drops that frame — documented, not crash.)

New tests, all TDD'd RED→GREEN: `TestRecoverPanicsDirIterator`, `TestRecoverPanicsFilesystemError` (Open/Stat/Remove), `TestConnDispatchRecoversPanic`, `TestConnDispatchPassesThroughError`.

Verification: `gofmt -l` clean, `go vet ./helpers/ .` clean, `go test -race ./helpers/ .` passes (incl. the full client↔server suite).

File: helpers/recovery.go [hunk 1]
```diff
@@ -12,9 +15,15 @@ import (
 // OnPanic is a function called when a panic is recovered.
 type OnPanic func(any)
 
+// ErrRecoveredPanic is returned by a billy.File or billy.Filesystem operation
+// that recovered from a panic, so a panicking call surfaces as an error to the
+// caller instead of an unwound stack or a zero-value "success".
```
Comment (arielshaqed) [inline helpers/recovery.go]:
```suggestion
// ErrRecoveredPanic is returned from a wrapped billy.File or billy.Filesystem operation
// that panics.
```

I do not believe it is worth spending words on what this does _not_ do.  The original has 14 words documenting the error (and can be shortened to 12), and another 21 explaining a bad alternative that you do not want to implement.

File: helpers/recovery.go [hunk 2]
```diff
@@ -12,9 +15,15 @@ import (
...
+var ErrRecoveredPanic = errors.New("recovered panic")
+
 // RecoverPanics wraps a handler to recover from panics in its method calls.
-// It also wraps any billy.Filesystem returned by the handler's Mount or FromHandle methods
-// to ensure panics in the filesystem implementation are caught.
+// It also wraps any billy.Filesystem returned by the handler's Mount or FromHandle
+// methods, and any billy.File those filesystems return, so panics in the
+// filesystem or file implementation are caught rather than crashing the server.
```
Comment (arielshaqed) [inline helpers/recovery.go]: This again spends more words explaining what would have happened if I did not use this.


```suggestion
// It also wraps any billy.Filesystem returned by the handler's Mount or FromHandle
// methods and any billy.File those filesystems return, to recover from panics and return them as errors.
```

File: conn_internal_test.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+// A panic in a handler (or in a billy operation that escaped the recovery
+// helpers) must not unwind the per-connection serve goroutine and crash the
+// process. dispatch recovers it and turns it into a SERVERFAULT so the request
+// fails with an NFS error reply instead.
+func TestConnDispatchRecoversPanic(t *testing.T) {
+	c := &conn{Server: &Server{}}
+	w := &response{conn: c, req: &request{}}
+
+	appError := c.dispatch(context.Background(), w, func(ctx context.Context, w *response, userHandler Handler) error {
```
Comment (arielshaqed) [inline conn_internal_test.go]: I don't understand what unit this is testing: isn't _any_ billy error supposed to return NFSStatusError?

**Review by arielshaqed — COMMENTED** [review]
Thanks!  This looks really neat, except for a confusing test and some overly long documentation describing what the code does _not_ d.

File: helpers/recovery.go [hunk 4]
```diff
@@ -12,9 +15,13 @@ import (
 // OnPanic is a function called when a panic is recovered.
 type OnPanic func(any)
 
+// ErrRecoveredPanic is returned from a wrapped billy.File or billy.Filesystem operation
+// that panics.
+var ErrRecoveredPanic = errors.New("recovered panic")
```
Comment (arielshaqed) [inline helpers/recovery.go]:
```suggestion
var ErrRecoveredPanic = errors.New("panic")
```
(the _error_ is that it panicked, not that it recovered...)

File: conn.go [hunk 5]
```diff
@@ -20,6 +21,8 @@ var (
 	ErrInputInvalid = errors.New("invalid input")
 	// ErrAlreadySent is returned when writing a header/status multiple times
 	ErrAlreadySent = errors.New("response already started")
+	// errHandlerPanic wraps a value recovered from a panic in a request handler.
+	errHandlerPanic = errors.New("recovered panic in handler")
```
Comment (arielshaqed) [inline conn.go]: same nit as before.

**Review by arielshaqed — APPROVED** [review]
Thanks!  Just 2 nits you may want to change.
